// react-control-library/shared/enumerations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationType = exports.PasswordStrength = void 0;
var PasswordStrength;
(function (PasswordStrength) {
    PasswordStrength["VeryStrong"] = "VeryStrong";
    PasswordStrength["Strong"] = "Strong";
    PasswordStrength["Good"] = "Good";
    PasswordStrength["Weak"] = "Weak";
    PasswordStrength["VeryWeak"] = "VeryWeak";
})(PasswordStrength = exports.PasswordStrength || (exports.PasswordStrength = {}));
var ValidationType;
(function (ValidationType) {
    ValidationType["maxLength"] = "maxLength";
    ValidationType["minLength"] = "minLength";
    ValidationType["min"] = "min";
    ValidationType["max"] = "max";
    ValidationType["userinput"] = "userInput";
    ValidationType["pattern"] = "pattern";
    ValidationType["required"] = "required";
    ValidationType["exactLength"] = "exactLength";
    ValidationType["strength"] = "strength";
    ValidationType["capital"] = "capital";
})(ValidationType = exports.ValidationType || (exports.ValidationType = {}));
